src/function/scalar/sequence/currval.cpp

```diff
@@ -0,0 +1,94 @@
+#include "duckdb/function/scalar/sequence_functions.hpp"
+
+#include "duckdb/catalog/catalog.hpp"
+#include "duckdb/catalog/catalog_entry/sequence_catalog_entry.hpp"
+#include "duckdb/common/exception.hpp"
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/execution/expression_executor.hpp"
+#include "duckdb/catalog/catalog.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+#include "duckdb/transaction/transaction.hpp"
+#include "duckdb/common/vector_operations/unary_executor.hpp"
+
+namespace duckdb {
+
+struct CurrvalBindData : public FunctionData {
+	//! The client context for the function call
+	ClientContext &context;
+	SequenceCatalogEntry *sequence;
+
+	CurrvalBindData(ClientContext &context, SequenceCatalogEntry *sequence) : context(context), sequence(sequence) {
+	}
+
+	unique_ptr<FunctionData> Copy() override {
+		return make_unique<CurrvalBindData>(context, sequence);
+	}
+};
+
+static int64_t CurrSequenceValue(Transaction &transaction, SequenceCatalogEntry *seq) {
+	lock_guard<mutex> seqlock(seq->lock);
+	int64_t result;
+	if (seq->usage_count == 0u) {
+		throw SequenceException("currval: sequence has not been used yet.");
+	}
+	result = seq->counter - seq->increment;
```

Comment:
Is the result not just `seq->counter`?

Reply:
Nevermind, currval should return the previously returned value of the sequence.

Reply:
The other option is not to pre-increment the counter in nextval, then just return the counter for currval.

Reply:
That is possible but you will have to take into account the other sequence options (start value, increment, minvalue, maxvalue, cycle, etc). I think in postgres they just keep a different value in the sequence object ("last value") that is set by calling nextval and read by currval.

Reply:
Actually I think the current implementation will fail for a cyclic sequence when the cycle resets and the sequence lines up the first value again. In that case subtracting it will result in the wrong answer. I will change it to Postgres implementation.

Reply:
Refactor is done